### DISCUSSION TITLE
Exclude files/folders from gazelle

### DISCUSSION THREAD

**Discussion by borjasantos-diez** [cgrindel/rules_swift_package_manager#678] [Q&A]
Hello, I'm having some issues with a few packages, where the BUILD file that gets generated is wrong.

As far as I can see, it's all about ignoring some folders/files, so I've been trying to provide that to `swift_update_packages` using the `kwargs`, but I haven't been able to make it work. I've also tried patching the repo, but one of the scenarios is not fixable that way, Any suggestions on how to use `kwargs` or any other way for gazelle to ignore stuff?

Thanks.

**Comment by cgrindel** [comment]
I would love to understand what is wrong so that we can fix it. Can you provide more information?

To answer your question, if you want gazelle to ignore certain paths, you can use the `# gazelle:exclude pattern`. Here is [an example](https://github.com/cgrindel/rules_swift_package_manager/blob/main/examples/firebase_example/BUILD.bazel#L15-L16) of how it is used.

**Reply by borjasantos-diez** [reply]
Thanks for the answer, @cgrindel. I'm already using `# gazelle:exclude xxx` for some other directories in my project, but in this case it doesn't seem to have effect in the generated files after checking out the dependencies through SPM. Should it just work if I exclude folders/files in the `BUILD.bazel` at the root of the project?

**Reply by cgrindel** [reply]
> Should it just work if I exclude folders/files in the BUILD.bazel at the root of the project?

Yes, if the pattern is correct.

Can you provide an example of what is being generated?

Also, just to clarify, the gazelle plugin only generates the build files for your source files. The build file generation for the external packages happens in Starlark code.

**Reply by borjasantos-diez** [reply]
Aha.. then I might be a bit mistaken :D First steps with Bazel in general, hehe. I'm getting errors like the following:
```
ERROR: /private/var/tmp/_bazel_borja/092ca09070fef0c5a810ca09a3a30c72/rules_xcodeproj.noindex/build_output_base/external/rules_swift_package_manager~0.16.0~swift_deps~swiftpkg_tagmanager_sdk_ios_public/BUILD.bazel:13:22: @rules_swift_package_manager~0.16.0~swift_deps~swiftpkg_tagmanager_sdk_ios_public//:sdk_resource_bundle: invalid label './DemoApp/ringieradsdk/Base.lproj/LaunchScreen.xib' in element 1 of attribute 'resources' in 'apple_resource_bundle' rule: invalid target name './DemoApp/ringieradsdk/Base.lproj/LaunchScreen.xib': target names may not contain '.' as a path segment
```
I'm having issues dealing with some dependencies that basically expose some `.xcframework`..

**Reply by cgrindel** [reply]
Can you provide a list of the dependencies? I can take a look at their package manifests.

**Reply by borjasantos-diez** [reply]
They are not public repositories, but I can probably attach their `Package.swift` if that helps in any way.. And one of the generated `BUILD.bazel` files
[Archive.zip](https://github.com/cgrindel/rules_swift_package_manager/files/13115368/Archive.zip)
[BUILD.bazel.zip](https://github.com/cgrindel/rules_swift_package_manager/files/13115526/BUILD.bazel.zip)

**Reply by cgrindel** [reply]
In the `RingierAd` package, the `sdk` target has an empty path. Can you try changing it to `"."` and see if that addresses the problem?

**Reply by cgrindel** [reply]
We may need to add a check for this in `rules_swift_package_manager`. However, I want to be sure that is the cause the leading `./` in the resource paths.

**Reply by borjasantos-diez** [reply]
Sure, I can just add a patch for that :) Will come back with an answer

**Reply by borjasantos-diez** [reply]
No changes after applying the patch, and I have just confirmed that both `Package.swift` now have `"."` :( 
```
ERROR: /private/var/tmp/_bazel_borja/092ca09070fef0c5a810ca09a3a30c72/rules_xcodeproj.noindex/build_output_base/external/rules_swift_package_manager~0.16.0~swift_deps~swiftpkg_tagmanager_sdk_xandr_public_bin/BUILD.bazel:19:22: @rules_swift_package_manager~0.16.0~swift_deps~swiftpkg_tagmanager_sdk_xandr_public_bin//:iAppNexusSDK_resource_bundle: invalid label './AppNexusSDK.xcframework/ios-arm64_x86_64-simulator/AppNexusSDK.framework/ANBrowserViewController.nib/objects-11.0+.nib' in element 0 of attribute 'resources' in 'apple_resource_bundle' rule: invalid target name './AppNexusSDK.xcframework/ios-arm64_x86_64-simulator/AppNexusSDK.framework/ANBrowserViewController.nib/objects-11.0+.nib': target names may not contain '.' as a path segment
```

**Reply by cgrindel** [reply]
I created #684 to track the issue.

**Reply by borjasantos-diez** [reply]
Will try to help as much as I can if you need me to. Many thanks.

**Reply by cgrindel** [reply]
I have [a PR](https://github.com/cgrindel/rules_swift_package_manager/pull/687) up that should address the bad resource paths.

**Reply by cgrindel** [reply]
The fix is released in [0.18.0](https://github.com/cgrindel/rules_swift_package_manager/releases/tag/v0.18.0). If you are using bzlmod, it will be available once [this PR](https://github.com/bazelbuild/bazel-central-registry/pull/1054) merges for the BCR.

**Reply by cgrindel** [reply]
There was a problem with pushing 0.18.0 to the BCR. So, if you are using bzlmod, use 0.18.1. Here is [the BCR PR](https://github.com/bazelbuild/bazel-central-registry/pull/1055).

**Reply by borjasantos-diez** [reply]
There also seem to be issues with 0.18.1, but don't worry. I'll keep track of the next release. No need to update in this discussion 👍 Many thanks for your efforts :)

**Comment by borjasantos-diez** [comment]
Hi @cgrindel,

I have just tested 0.18.2, and even though the issue with the `./` is fixed, I'm still getting some errors, due to some files with the same name being added to the resources bundle
```
Error in fail: Multiple files would be placed at "iAppNexusSDK_resource_bundle.bundle/objects-11.0+.nib" in the bundle, which is not allowed. check input file:
external/rules_swift_package_manager~0.18.2~swift_deps~swiftpkg_tagmanager_sdk_xandr_public_bin/AppNexusSDK.xcframework/ios-arm64_x86_64-simulator/AppNexusSDK.framework/ANInterstitialAdViewController.nib/objects-11.0+.nib
```
Would it make sense to use the `structured_resources` attribute rather than `resources` in this case so that things are kept in its place?

I've attached the generated `BUILD.bazel`
[BUILD.bazel.zip](https://github.com/cgrindel/rules_swift_package_manager/files/13187289/BUILD.bazel.zip)

**Reply by cgrindel** [reply]
From my read of [the SPM code](https://github.com/apple/swift-package-manager/blob/main/Sources/PackageModel/Resource.swift#L25-L33), it looks like SPM does not support structured resources. 🤔

**Reply by luispadron** [reply]
I do think they copy to a given structure, under some instances:

> Copy rule
> Some Swift packages may require a resource file to remain untouched or to retain a certain directory structure for resources. Use the [copy(_:)](https://developer.apple.com/documentation/PackageDescription/Resource/copy(_:)) function to apply this rule and have Xcode copy the resource as is to the top level of the resource bundle. If you pass a directory path to the copy rule, Xcode retains the directory’s structure.
>
https://developer.apple.com/documentation/xcode/bundling-resources-with-a-swift-package#Explicitly-declare-or-exclude-resources

**Reply by cgrindel** [reply]
@borjasantos-diez Is there any chance that you could share the `Package.swift` that was used to generated the `BUILD.bazel` file? I am worried that `rules_swift_package_manager` should not have created the `apple_resource_bundle` in this case.

**Reply by borjasantos-diez** [reply]
Sure, I've already shared it in the conversation.
https://github.com/cgrindel/rules_swift_package_manager/files/13115368/Archive.zip
That zip file contains a folder called `tagmanager-sdk-xandr-public-bin` with the `Package.swift`. It's a package just exposing a couple of `.xcframework`.

**Reply by cgrindel** [reply]
The `apple_resource_bundle` named `iAppNexusSDK_resource_bundle` seems to only include files from the `AppNexusSDK` xcframework. I believe that this is happening because the `iAppNexusSDK` target is not excluding the xcframework paths. Can you try excluding those paths from the `iAppNexusSDK` target?

```swift
    targets: [
        .target(name: sdkName, path: ""),   // <=== I think this should exclude AppNexusSDK.xcframework and OMSDK_Appnexus.xcframework.
        .binaryTarget(name: "AppNexusSDK", path: "AppNexusSDK.xcframework"),
        .binaryTarget(name: "OMSDK_Appnexus", path: "OMSDK_Appnexus.xcframework"),
    ],
```

**Reply by borjasantos-diez** [reply]
I've added a patch, but I'm still getting the same error... I've attached the full folder that gets generated, with the modified `Package.swift`, in case that provides any more info.
[rules_swift_package_manager~0.18.2~swift_deps~swiftpkg_tagmanager_sdk_xandr_public_bin.zip](https://github.com/cgrindel/rules_swift_package_manager/files/13217646/rules_swift_package_manager.0.18.2.swift_deps.swiftpkg_tagmanager_sdk_xandr_public_bin.zip)

**Reply by cgrindel** [reply]
OK. I think that I may understand what is happening. The path exclusions were not being passed to the resource discovery functions. I put together [a PR](https://github.com/cgrindel/rules_swift_package_manager/pull/709) that attempts to fix this. Can you try using the branch with [git_override](https://bazel.build/rules/lib/globals/module#git_override)?

Add something like the following to your `MODULE.bazel`:
```python
git_override(
    module_name = "rules_swift_package_manager",
    remote = "https://github.com/cgrindel/rules_swift_package_manager",
    commit = "57277302deb455cceb27bf84e4a56a41871d0fbd",  # commit for branch https://github.com/cgrindel/rules_swift_package_manager/tree/gh678_res_bundle_disc
)
```

**Reply by cgrindel** [reply]
I made some updates to the branch. Here is the new commit: `ec4b89256e283bb67aa3f29b73a582bfe4eede98`.

**Reply by borjasantos-diez** [reply]
Brilliant! It has finally worked :) I still have to patch the `Package.swift` but that's something I can send some feedback about, or even create a PR for them to add those excludes.
Many thanks for your time 🙏

**Comment by DJBen** [comment]
I too need to exclude path. Use case: I have a integration test subdirectory that has its own `MODULE.bazel`. I don't want my toplevel gazelle update to affect that.

**Reply by cgrindel** [reply]
Have you tried using the `# gazelle:exclude pattern` directive?
